Yosemite/Yosemite/Stores/Helpers/CouponService.swift

```diff
@@ -0,0 +1,135 @@
+    ///   - pageNumber: Page number of coupons to fetch from the API
+    ///   - pageSize: Number of coupons per page to fetch from the API
+    ///   - onCompletion: Closure to call after sychronizing is complete. Called on the main thread.
+    ///   - result: `.success(hasNextPage: Bool)` or `.failure(error: Error)`
+    ///
+    func synchronizeCoupons(siteID: Int64,
+                            pageNumber: Int,
+                            pageSize: Int,
+                            onCompletion: @escaping (_ result: Result<Bool, Error>) -> Void) {
+        remote.loadAllCoupons(for: siteID,
+                              pageNumber: pageNumber,
+                              pageSize: pageSize) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .failure(let error):
+                onCompletion(.failure(error))
+
+            case .success(let coupons):
+                let shouldClearData = pageNumber == Remote.Default.firstPageNumber
+                let hasNextPage = coupons.count == pageSize
+                self.upsertStoredCouponsInBackground(readOnlyCoupons: coupons,
+                                                     siteID: siteID,
+                                                     shouldClearExistingCoupons: shouldClearData) {
+                    onCompletion(.success(hasNextPage))
+                }
+            }
+        }
+    }
+
+    // MARK: - Storage Coupon
+
+    /// Updates or Inserts specified Coupon Entities in a background thread
+    /// `onCompletion` will be called on the main thread.
+    ///
+    func upsertStoredCouponsInBackground(readOnlyCoupons: [Networking.Coupon],
+                                         siteID: Int64,
+                                         shouldClearExistingCoupons: Bool = false,
+                                         onCompletion: @escaping () -> Void) {
+        storageManager.performAndSave({ [weak self] storage in
+            guard let self else { return }
+            if shouldClearExistingCoupons {
+                storage.deleteCoupons(siteID: siteID)
+            }
+            upsertStoredCoupons(readOnlyCoupons: readOnlyCoupons,
+                                in: storage,
+                                siteID: siteID)
+        }, completion: onCompletion, on: .main)
+    }
+
+    /// Updates or Inserts the specified Coupon entities
+    ///
+    func deleteStoredCoupon(siteID: Int64, couponID: Int64, onCompletion: @escaping () -> Void) {
+        storageManager.performAndSave({ storage in
+            storage.deleteCoupon(siteID: siteID, couponID: couponID)
+        }, completion: onCompletion, on: .main)
+    }
+
+    /// Deletes the Storage.Coupon with the specified `siteID` and `couponID` in a background thread.
+    /// Triggers `onCompletion` on the main thread when done.
+    ///
+    func upsertStoredCoupons(readOnlyCoupons: [Networking.Coupon],
```

Comment:
This one it's only called here, so we can restrict it further:
```suggestion
    private func upsertStoredCoupons(readOnlyCoupons: [Networking.Coupon],
```